\documentclass[]{elsarticle} %review=doublespace preprint=single 5p=2 column
%%% Begin My package additions %%%%%%%%%%%%%%%%%%%
\usepackage[hyphens]{url}

  \journal{An awesome journal} % Sets Journal name


\usepackage{lineno} % add
\providecommand{\tightlist}{%
  \setlength{\itemsep}{0pt}\setlength{\parskip}{0pt}}

\usepackage{graphicx}
\usepackage{booktabs} % book-quality tables
%%%%%%%%%%%%%%%% end my additions to header

\usepackage[T1]{fontenc}
\usepackage{lmodern}
\usepackage{amssymb,amsmath}
\usepackage{ifxetex,ifluatex}
\usepackage{fixltx2e} % provides \textsubscript
% use upquote if available, for straight quotes in verbatim environments
\IfFileExists{upquote.sty}{\usepackage{upquote}}{}
\ifnum 0\ifxetex 1\fi\ifluatex 1\fi=0 % if pdftex
  \usepackage[utf8]{inputenc}
\else % if luatex or xelatex
  \usepackage{fontspec}
  \ifxetex
    \usepackage{xltxtra,xunicode}
  \fi
  \defaultfontfeatures{Mapping=tex-text,Scale=MatchLowercase}
  \newcommand{\euro}{â‚¬}
\fi
% use microtype if available
\IfFileExists{microtype.sty}{\usepackage{microtype}}{}
\bibliographystyle{elsarticle-harv}
\ifxetex
  \usepackage[setpagesize=false, % page size defined by xetex
              unicode=false, % unicode breaks when used with xetex
              xetex]{hyperref}
\else
  \usepackage[unicode=true]{hyperref}
\fi
\hypersetup{breaklinks=true,
            bookmarks=true,
            pdfauthor={},
            pdftitle={Short Paper},
            colorlinks=false,
            urlcolor=blue,
            linkcolor=magenta,
            pdfborder={0 0 0}}
\urlstyle{same}  % don't use monospace font for urls

\setcounter{secnumdepth}{0}
% Pandoc toggle for numbering sections (defaults to be off)
\setcounter{secnumdepth}{0}

% Pandoc citation processing
\newlength{\csllabelwidth}
\setlength{\csllabelwidth}{3em}
\newlength{\cslhangindent}
\setlength{\cslhangindent}{1.5em}
% for Pandoc 2.8 to 2.10.1
\newenvironment{cslreferences}%
  {}%
  {\par}
% For Pandoc 2.11+
\newenvironment{CSLReferences}[3] % #1 hanging-ident, #2 entry spacing
 {% don't indent paragraphs
  \setlength{\parindent}{0pt}
  % turn on hanging indent if param 1 is 1
  \ifodd #1 \everypar{\setlength{\hangindent}{\cslhangindent}}\ignorespaces\fi
  % set entry spacing
  \ifnum #2 > 0
  \setlength{\parskip}{#2\baselineskip}
  \fi
 }%
 {}
\usepackage{calc} % for calculating minipage widths
\newcommand{\CSLBlock}[1]{#1\hfill\break}
\newcommand{\CSLLeftMargin}[1]{\parbox[t]{\csllabelwidth}{#1}}
\newcommand{\CSLRightInline}[1]{\parbox[t]{\linewidth - \csllabelwidth}{#1}}
\newcommand{\CSLIndent}[1]{\hspace{\cslhangindent}#1}

% Pandoc header



\begin{document}
\begin{frontmatter}

  \title{Short Paper}
    \author[University at Albany]{Elliot Todd Martin\corref{1}}
   \ead{etmartin@albany.edu} 
    \author[University at Albany,BMS]{Noor Kotb}
  
    \author[University at Albany]{Gabriele Fuchs}
  
    \author[University at Albany]{Prashanth Rangan\corref{1}}
   \ead{prangan@albany.edu} 
      \address[University at Albany]{Department of Biological
Sciences/RNA Institute, University at Albany SUNY, Albany, NY 12202,
USA.}
    \address[BMS]{Department of Biomedical Sciences, School of Public
Health, University at Albany SUNY, Rensselaer, NY, 12144, USA.}
      \cortext[1]{Corresponding Author}
    \cortext[2]{Equal contribution}
  
  \begin{abstract}
  Proper stem cell differentiation requires both transcriptional and
  post-transcriptional changes. Previous work at characterizing these
  changes have primarily focused either on cell culture systems, which
  lack biological context. In-vivo work on stem cell systems has
  biological context, but these models are less tractable to high
  throughput methods especially for post-transcriptional changes. Here,
  we have compiled and developed tooling for mRNA level and
  post-transcriptional data that represent several stages of
  differentiation of the female \emph{Drosophila} germline stem cell
  (GSC) differentiation program. We have developed visualization tools
  to make this data accessible to non-bioinformaticians that are
  accessible through a browser. We confirmed the expression of genes
  that have been described previously and elucidate\ldots{}
  \end{abstract}
  
 \end{frontmatter}

\hypertarget{introduction}{%
\section{Introduction:}\label{introduction}}

The female \emph{Drosophila} germline provides a powerful system to
study stem cell differentiation in an in-vivo setting. \emph{Drosophila}
are one of the most genetically tractable model organisms available and
hundreds of thousands of mutant, RNAi, overexpression, and reporter
lines are commercially available. Additionally, each of stage of
differentiation of \emph{Drosophila} female germline stem cell (GSCs)
are observable and identifiable from a single ovary allowing for
temporal changes over GSC development to be easily studied. However, one
weakness to studying stem cell differentiation within a tissue has been
the inaccessibility to high throughput methods that typically require a
relatively homogenous collection of cells. Single cell seq promises to
bridge this gap however, it has several limitations which have not been
overcome as of the writing of this manuscript. First, single cell-seq is
limited to mRNA level data, it cannot yet be applied to techniques such
as polysome or ribosome sequencing. Second, it has a higher limit of
detection than bulk mRNA-seq, preventing the detection of rare, but
potentially important mRNAs. Here, we use \emph{Drosophila} genetics in
order to circumvent the limitations of single cell seq and apply bulk
sequencing techniques to tissue enriched for several stages of GSC
differentiation. We present this data, alongside previously published
single cell-seq data from \emph{Drosophila} ovaries in a tool called
Ovary-App.

\hypertarget{results}{%
\section{Results:}\label{results}}

\hypertarget{ovary-app}{%
\paragraph{Ovary-App}\label{ovary-app}}

\hfill\break

We present the tool Ovary-App which consists of a collection of
user-interactable visualizations allowing researchers to easily
determine the expression pattern of a gene of interest or the expression
pattern of a collection of genes provided by the user. Ovary-App
consists of three modules, ovary-map, ovary-heatmap, and ovary-violin.
Each module of Ovary-App allows users to visualize expression from
matched mRNA seq and polysome-seq data of genetically enriched stages of
early GSC differentiation as well as previously published single-seq
data. Ovary-map allows users to visualize the expression of a single
gene over the course of differentiation in the form of a cartoon
germarium, which eases understanding of staging for those less familiar
with \emph{Drosophila} oogenesis. Ovary-heatmap consists of a clustered,
interactive heatmap that allows users to explore expression trends over
development. Finally, ovary-violin allows users to visualize the
expression of multiple genes over the course of differentiation. These
groups of genes can be selected either by a GO-term of interest or a
custom list of genes can be supplied by the user. Researchers can use
these datasets to enhance their hypothesis generation or to confirm
expression patterns they have observed from other methods.

\hypertarget{mrna-expression-validation}{%
\paragraph{mRNA Expression
Validation}\label{mrna-expression-validation}}

\hfill\break

To determine if the bulk mRNA-seq data we have collected is
representative of previously observed expression patterns, we first
compared our mRNA-seq data to previously reported mRNA-seq data
generated from similar genetic enrichment strategies to enrich for GSCs
and GSC-daughter cells, however, the previously published data included
a FACS step so that only are pure population of germline cells were
sequenced. Indeed, we find that genes identified as being differentially
expressed from the previously published data follow similar trends in
our data, indicating that despite the lack of FACS our data reproduces
meaningful expression changes at the mRNA level. To validate our mRNA
seq data for the genetically enriched stages for which no previous mRNA
seq libraries have been published and to ensure that the mRNA seq
results we observe recapitulate we used in-situ probes targeting RpS19b.
Our mRNA-seq data as well as the available SC-seq data indicates that
RpS19b is highly expressed in GSCs that decreases over differentiation
with greatly decreased expression in egg chambers (Fig n).

\hypertarget{translation-efficiency-validation}{%
\paragraph{Translation Efficiency
Validation}\label{translation-efficiency-validation}}

\hfill\break

Next, to determine if the polysome-seq data is representative of
biologically meaningful changes in translation status, we (either do
staining/in situ or refer) examined the expression pattern of a gene
that is known to have dynamic expression during GSC differentiation. We
chose to examine Bru1 because it has been previously observed to
dramatically increase in expression during the cyst stages of GSC
differentiation at the protein level. Our Ovary-App data suggests that
Bru1 mRNA levels are relatively consistent during early oogenesis, both
from bulk mRNA-seq (Fig n) and SC-seq (Fig n), until the egg chamber
stages where Bru1 levels dramatically increase. However, our
polysome-seq data is consistent with the observation that Bru1
expression increases during the cyst stages, at the level of
translation. This led us to predict that Bru1 mRNA levels would be
consistent GSC differentiation, but the protein expression would
increase during the cyst stages, implying a change in the translation
status of Bru1 mRNA. Another mRNA whose translation is well
characterized and tightly regulated during GSC differentiation is pgc.
Pgc is a maternally deposited mRNA that encodes a protein that silences
RNA Pol II via phosphorylation of the Ser2 residue of Pol II. During GSC
differentiation, pgc mRNA is expressed, but only translated in a short
burst in stem cell daughters, prior to Bam expression. Interestingly,
from our mRNA-seq data, we observe a dramatic dip in the mRNA level of
pgc in stem cell daughters and a simultaneous increase in the
translation efficiency. Although the reason for the lowered mRNA
expression in stem cell daughter cells is unknown, the increase in pgc
translation is consistent with its reported expression.

\hypertarget{bibliography-styles}{%
\section{Bibliography styles}\label{bibliography-styles}}

There are various bibliography styles available. You can select the
style of your choice in the preamble of this document. These styles are
Elsevier styles based on standard styles like Harvard and Vancouver.
Please use BibTeX~to generate your bibliography and include DOIs
whenever available.

Here are two sample references: (Dirac, 1953; Feynman and Vernon Jr.;
1963).

\hypertarget{references}{%
\section*{References}\label{references}}
\addcontentsline{toc}{section}{References}

\hypertarget{refs}{}
\begin{CSLReferences}{1}{0}
\leavevmode\hypertarget{ref-Dirac1953888}{}%
Dirac, P.A.M., 1953. The lorentz transformation and absolute time.
Physica 19, 888--896.
doi:\href{https://doi.org/10.1016/S0031-8914(53)80099-6}{10.1016/S0031-8914(53)80099-6}

\leavevmode\hypertarget{ref-Feynman1963118}{}%
Feynman, R.P., Vernon Jr., F.L., 1963. The theory of a general quantum
system interacting with a linear dissipative system. Annals of Physics
24, 118--173.
doi:\href{https://doi.org/10.1016/0003-4916(63)90068-X}{10.1016/0003-4916(63)90068-X}

\end{CSLReferences}


\end{document}

